Ugh, and now a mock that works on F25 doesn't work on EL 7:

```
AssertionError: Expected call: listTagged('f26-updates-testing', latest=True) Actual call: listTagged("f<MagicMock name='open().readlines().__getitem__().split().__getitem__()' id='83276240'>-updates-testing", latest=True)
```

--- a/bodhi/client/bindings.py
+++ b/bodhi/client/bindings.py
@@ -36,6 +36,7 @@ import re
 import textwrap
 
 from six.moves import configparser
+import dnf
 import six
 
 from fedora.client import AuthError, OpenIdBaseClient, FedoraClientError
@@ -447,7 +448,6 @@ class BodhiClient(OpenIdBaseClient):
 
         Only works on systems with dnf.
         """
-        import dnf
         base = dnf.Base()
         sack = base.fill_sack(load_system_repo=True)
         query = sack.query()

--- a/bodhi/client/bindings.py
+++ b/bodhi/client/bindings.py
@@ -580,7 +580,7 @@ class BodhiClient(OpenIdBaseClient):
                 {"dist_tag": "dist-f12", "id_prefix": "FEDORA",
                  "locked": false, "name": "F12", "long_name": "Fedora 12"}]}
         """
-        return self.send_request('releases/', params=kwargs)
+        return self.send_request('releases/', verb='GET', params=kwargs)
 
     def get_koji_session(self):
         """ Return an authenticated koji session """

--- a/bodhi/client/bindings.py
+++ b/bodhi/client/bindings.py
@@ -35,8 +35,10 @@ import os
 import re
 import textwrap
 
+from iniparse.compat import ConfigParser
 from six.moves import configparser
 import dnf
+import koji
 import six
 
 from fedora.client import AuthError, OpenIdBaseClient, FedoraClientError
@@ -455,8 +457,7 @@ class BodhiClient(OpenIdBaseClient):
         with open('/etc/fedora-release', 'r') as f:
             fedora = f.readlines()[0].split()[2]
         tag = 'f%s-updates-testing' % fedora
-        builds = self.get_koji_session(
-            login=False).listTagged(tag, latest=True)
+        builds = self.get_koji_session().listTagged(tag, latest=True)
         for build in builds:
             pkgs = installed.filter(name=build['name'], version=build['version'],
                                     release=build['release']).run()
@@ -581,21 +582,14 @@ class BodhiClient(OpenIdBaseClient):
         """
         return self.send_request('releases/', params=kwargs)
 
-    def get_koji_session(self, login=True):
+    def get_koji_session(self):
         """ Return an authenticated koji session """
-        import koji
-        from iniparse.compat import ConfigParser
         config = ConfigParser()
         if os.path.exists(os.path.join(os.path.expanduser('~'), '.koji', 'config')):
             config.readfp(open(os.path.join(os.path.expanduser('~'), '.koji', 'config')))
         else:
             config.readfp(open('/etc/koji.conf'))
-        cert = os.path.expanduser(config.get('koji', 'cert'))
-        ca = os.path.expanduser(config.get('koji', 'ca'))
-        serverca = os.path.expanduser(config.get('koji', 'serverca'))
         session = koji.ClientSession(config.get('koji', 'server'))
-        if login:
-            session.ssl_login(cert=cert, ca=ca, serverca=serverca)
         return session
 
     koji_session = property(fget=get_koji_session)
@@ -610,7 +604,7 @@ class BodhiClient(OpenIdBaseClient):
             raise BodhiClientException('You must specify a username')
         builds = []
         data = self.get_releases()
-        koji = self.get_koji_session(login=False)
+        koji = self.get_koji_session()
         for release in data['releases']:
             try:
                 for build in koji.listTagged(release['candidate_tag'], latest=True):

--- a/bodhi/tests/client/test_bindings.py
+++ b/bodhi/tests/client/test_bindings.py
@@ -1057,6 +1057,29 @@ class TestUpdateNotFound(unittest.TestCase):
         self.assertEqual(unicode(exc), 'Update not found: bodhi-2.2.4-1.el7')
 
 
+class TestBodhiClient_get_releases(unittest.TestCase):
+    """
+    Test the BodhiClient.get_releases() method.
+    """
+    @mock.patch('bodhi.client.bindings.BodhiClient._load_cookies', mock.MagicMock())
+    def test_get_releases(self):
+        """Assert correct behavior from the get_releases() method."""
+        client = bindings.BodhiClient()
+        client.send_request = mock.MagicMock(
+            return_value={'releases': [{'candidate_tag': 'f25-updates-testing'},
+                                       {'candidate_tag': 'f26-updates-testing'}]})
+
+        results = client.get_releases(some_param='some_value')
+
+        self.assertEqual(
+            results,
+            {'releases': [
+                {'candidate_tag': 'f25-updates-testing'},
+                {'candidate_tag': 'f26-updates-testing'}]})
+        client.send_request.assert_called_once_with(
+            'releases/', params={'some_param': 'some_value'}, verb='GET')
+
+
 class TestBodhiClient_parse_file(unittest.TestCase):
     """
     Test the BodhiClient.parse_file() method.

--- a/bodhi/tests/client/test_bindings.py
+++ b/bodhi/tests/client/test_bindings.py
@@ -1142,3 +1142,39 @@ class TestBodhiClient_parse_file(unittest.TestCase):
 
         self.assertEqual(unicode(exc.exception),
                          'No such file or directory: /tmp/bodhi-test-parsefile2')
+
+
+class TestBodhiClient_testable(unittest.TestCase):
+    """
+    Test the BodhiClient.testable() method.
+    """
+    @mock.patch('__builtin__.open', create=True)
+    @mock.patch('bodhi.client.bindings.BodhiClient._load_cookies', mock.MagicMock())
+    @mock.patch('bodhi.client.bindings.BodhiClient.get_koji_session')
+    @mock.patch('bodhi.client.bindings.dnf.Base.fill_sack')
+    def test_testable(self, fill_sack, get_koji_session, mock_open):
+        """Assert correct behavior from the testable() method."""
+        get_koji_session.return_value.listTagged.return_value = [
+            {'name': 'bodhi', 'version': '2.9.0', 'release': '1.fc26', 'nvr': 'bodhi-2.9.0-1.fc26'}]
+        fill_sack.return_value.query.return_value.installed.return_value.filter.return_value.run.\
+            return_value = ['bodhi-2.8.1-1.fc26']
+        mock_open.return_value.__enter__.return_value.readlines.return_value = [
+            'Fedora release 26 (Twenty Six)']
+        client = bindings.BodhiClient()
+        client.send_request = mock.MagicMock(
+            return_value={'updates': [{'nvr': 'bodhi-2.9.0-1.fc26'}]})
+
+        updates = client.testable()
+
+        self.assertEqual(list(updates), [{'nvr': 'bodhi-2.9.0-1.fc26'}])
+        fill_sack.assert_called_once_with(load_system_repo=True)
+        fill_sack.return_value.query.assert_called_once_with()
+        fill_sack.return_value.query.return_value.installed.assert_called_once_with()
+        fill_sack.return_value.query.return_value.installed.return_value.filter.\
+            assert_called_once_with(name='bodhi', version='2.9.0', release='1.fc26')
+        fill_sack.return_value.query.return_value.installed.return_value.filter.return_value.run.\
+            assert_called_once_with()
+        get_koji_session.return_value.listTagged.assert_called_once_with('f26-updates-testing',
+                                                                         latest=True)
+        client.send_request.assert_called_once_with(
+            'updates/', params={'builds': 'bodhi-2.9.0-1.fc26'}, verb='GET')

--- a/devel/ansible/roles/dev/tasks/main.yml
+++ b/devel/ansible/roles/dev/tasks/main.yml
@@ -58,6 +58,7 @@
       - python2-colander
       - python2-createrepo_c
       - python2-cryptography
+      - python2-dnf
       - python2-fedmsg-atomic-composer
       - python2-fedmsg-commands
       - python2-fedmsg-consumers

--- a/devel/ci/run_tests.sh
+++ b/devel/ci/run_tests.sh
@@ -31,6 +31,7 @@ sudo yum install -y\
     python-cornice\
     python-createrepo_c\
     python-cryptography\
+    python-dnf\
     python-dogpile-cache\
     python-fedora\
     python-kitchen\

--- a/docs/release_notes.rst
+++ b/docs/release_notes.rst
@@ -1,6 +1,18 @@
 Release notes
 =============
 
+develop
+-------
+
+Dependency changes
+------------------
+
+dnf is now a required dependency for the Python bindings. Technically, it was needed before for some
+of the functionality, and it would traceback if that functionality was used without dnf being
+present. With this change, the module will fail to import without dnf, and dnf is a formal
+dependency.
+
+
 2.9.0
 -----
 

